node/modules/storageminer.go

```diff
@@ -357,78 +357,193 @@ func SectorStorage(mctx helpers.MetricsCtx, lc fx.Lifecycle, lstor *paths.Local,
-				log.Errorf("Cannot participate in F3 as it is disabled: %+v", err)
-				return
-			}
-			if err != nil {
-				log.Errorf("while starting to participate in F3: %+v", err)
-				// use exponential backoff to avoid hotloop
-				timer.Reset(b.Duration())
-				select {
-				case <-ctx.Done():
-					return
-				case <-timer.C:
-				}
-				continue
-			}
-			if !ok {
-				log.Errorf("lotus node refused our lease, are you loadbalancing or did the miner just restart?")
-
-				sleepFor := b.Duration()
-				if d := time.Until(oldLease); d > 0 && d < sleepFor {
-					sleepFor = d
-				}
-				timer.Reset(sleepFor)
-				select {
-				case <-ctx.Done():
-					return
-				case <-timer.C:
-				}
-				continue
+func (p *f3Participator) awaitLeaseExpiry(ctx context.Context, lease api.F3ParticipationLease) error {
+	p.backoff.Reset()
+	for ctx.Err() == nil {
+		switch progress, err := p.node.F3GetProgress(ctx); {
+		case errors.Is(err, api.ErrF3Disabled):
+			log.Errorw("Cannot await F3 participation lease expiry as F3 is disabled.", "err", err)
+			return xerrors.Errorf("awaiting F3 participation lease expiry: %w", err)
+		case err != nil:
+			if p.backoff.Attempt() > float64(p.maxCheckProgressAttempts) {
+				log.Errorw("Too many failures while attempting to check F3  progress. Restarting participation.", "attempts", p.backoff.Attempt(), "err", err)
+				return nil
 			}
+			log.Errorw("Failed to check F3 progress while awaiting lease expiry. Retrying after backoff.", "attempts", p.backoff.Attempt(), "backoff", p.backoff.Duration(), "err", err)
+			p.backOff(ctx)
+		case progress.Instance+2 >= lease.ValidityTerm:
+			log.Infof("F3 progressed (%d) to within two instances of lease expiry (%d). Restarting participation.", progress.Instance, lease.ValidityTerm)
+			return nil
+		default:
+			remainingInstanceLease := lease.ValidityTerm - progress.Instance
+			log.Debugf("F3 participation lease is valid for further %d instances. Recking after %s.", remainingInstanceLease, p.checkProgressInterval)
+			p.backOffFor(ctx, p.checkProgressInterval)
+		}
+	}
+	return ctx.Err()
+}
 
-			// we have succeeded in giving a lease, reset the backoff
-			b.Reset()
+func (p *f3Participator) backOff(ctx context.Context) {
+	p.backOffFor(ctx, p.backoff.Duration())
+}
 
-			oldLease = newLease
-			// wait for the half of the lease time and then refresh
-			timer.Reset(leaseTime / 2)
-			select {
-			case <-ctx.Done():
-				return
-			case <-timer.C:
-			}
+func (p *f3Participator) backOffFor(ctx context.Context, d time.Duration) {
+	// Create a timer every time to avoid potential risk of deadlock or the need for
+	// mutex despite the fact that f3Participator is never (and should never) be
+	// called from multiple goroutines.
+	timer := time.NewTimer(d)
+	select {
+	case <-ctx.Done():
+		return
+	case <-timer.C:
+	}
+}
+
+func F3Participation(mctx helpers.MetricsCtx, lc fx.Lifecycle, node v1api.FullNode, participant dtypes.MinerAddress) error {
```

Comment:
Can we use this in the itest?

Reply:
I _think_ [we already do](https://github.com/filecoin-project/lotus/blob/63595bb73df0b673eeaa31a0d7dcba7d2823d077/itests/f3_test.go#L192)?

Reply:
Oh, I see. I thought this was new.

Reply:
The `f3Participator` construct is new. it refactors the logic that was in a loop inside `F3Participation` into its own type.